framework/src/base/Assembly.C

```diff
@@ -1389,3 +1389,43 @@ Assembly::addJacobianOffDiagScalar(SparseMatrix<Number> & jacobian, unsigned int
       addJacobianBlock(jacobian, jacobianBlock(var_i.number(), var_j.number()), var_i.dofIndices(), var_j.dofIndices(), var_i.scalingFactor());
   }
 }
+
+void
+Assembly::cacheNodalBCJacobianEntry(numeric_index_type i, numeric_index_type j, Real value)
+{
+  _cached_nodal_bc_rows.push_back(i);
+  _cached_nodal_bc_cols.push_back(j);
+  _cached_nodal_bc_vals.push_back(value);
+}
+
+void
+Assembly::clearCachedNodalBCJacobianEntries()
+{
+  unsigned int orig_capacity = _cached_nodal_bc_rows.size();
+
+  _cached_nodal_bc_rows.clear();
+  _cached_nodal_bc_cols.clear();
+  _cached_nodal_bc_vals.clear();
+
+  // It's possible (though massively unlikely) that clear() will
+  // change the capacity of the vectors, so let's be paranoid and
+  // explicitly reserve() the same amount of memory to avoid multiple
+  // push_back() induced allocations.
+  _cached_nodal_bc_rows.reserve(orig_capacity);
```

Comment:
I think that would be fine.  This is at the level of micro-optimization at this point... so without a real use/test case there's no reason to go any further...